packages/preview/src/browser/link-normalizer.ts

```diff
@@ -0,0 +1,26 @@
+/********************************************************************************
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import URI from '@theia/core/lib/common/uri';
+
+@injectable()
+export class LinkNormalizer {
+
+    normalizeLink(documentUri: URI, link: string): string {
+        return 'mini-browser/' + documentUri.parent.resolve(link).path.toString();
```

Comment:
How about adding a dependency to it?

Reply:
to mini-browser?
not sure, maybe other apps want to rebind this normalizer and use another endpoint.

WDYT?

Reply:
I will just add a dependency.

If there should be another implementation I would split the preview extension on api and impl extensions, then the minibrowse can depend on the api part and contribute `MiniBrowserPreviewLinkNormalizer` without registering anything by the impl extension.

Reply:
+1 for introducing it.

Hopefully, we could reuse the [`LocationMapperService`](https://github.com/theia-ide/theia/blob/40640d5cca03c82536d4c371eb962e20a7f9097c/packages/mini-browser/src/browser/location-mapper-service.ts#L47). It can handle `file`, `http`, and `https` URI schemes.

Let me know if I can help with something in this.

Reply:
> +1 for introducing it.

Cool. Which one? ;-)

1. `preview --> mini-browser`; preview just uses mini-browser api
2. `mini-browser --> preview`; mini-browser rebinds default impl for link normalizer
3. extract `md` out of `preview`, which would remain as base and api; `mini-browser` would contribute the normalizer

Should we start simple? Though, I do not like such hard dependencies in 1. or 2.

OTOH we could create `@theia/markdown` now and add LS later? 

> Hopefully, we could reuse the LocationMapperService. It can handle file, http, and https URI schemes.

Thanks @kittaakos, it's a good idea to pass the location through this to get the final URL. Catching up with changes is silly.

Reply:
> Should we start simple? Though, I do not like such hard dependencies in 1. or 2.

+1 for the simple way. The first one is straightforward. But it's your call.

Reply:
call site is not async, so I was happy about the MaybePromise.